tests/regression_tests/trigger_statepoint_restart/test.py

```diff
@@ -0,0 +1,132 @@
+import glob
+import os
+
+import openmc
+import pytest
+
+from tests.testing_harness import PyAPITestHarness
+from tests.regression_tests import config
+
+
+@pytest.fixture
+def model():
+
+    # Materials
+    materials = openmc.Materials()
+    mat = openmc.Material()
+    mat.set_density('g/cm3', 4.5)
+    mat.add_nuclide('U235', 1.0)
+    materials.append(mat)
+
+    # Geometry
+    sph = openmc.Sphere(r=10.0, boundary_type='vacuum')
+    cell = openmc.Cell(fill=mat, region=-sph)
+    geometry = openmc.Geometry([cell])
+
+    # Settings
+    settings = openmc.Settings()
+    settings.run_mode = 'eigenvalue'
+    settings.batches = 10
+    settings.inactive = 5
+    settings.particles = 200
+    settings.keff_trigger = {'type': 'std_dev', 'threshold': 0.002}
```

Comment:
I noticed that on Python 3.5, this test fails due to the inputs not matching. The underlying reason is that dictionaries in Python 3.5 are not guaranteed to retain insertion order. As a remedy, what we can do is sort the (key, value) pairs coming out of the dictionary when we write them to the settings.xml file:
```diff
diff --git a/openmc/settings.py b/openmc/settings.py
index 794c97b61..56125365e 100644
--- a/openmc/settings.py
+++ b/openmc/settings.py
@@ -809,9 +809,9 @@ class Settings(object):
         if self._keff_trigger is not None:
             element = ET.SubElement(root, "keff_trigger")
 
-            for key in self._keff_trigger:
+            for key, value in sorted(self._keff_trigger.items()):
                 subelement = ET.SubElement(element, key)
-                subelement.text = str(self._keff_trigger[key]).lower()
+                subelement.text = str(value).lower()
 
     def _create_energy_mode_subelement(self, root):
         if self._energy_mode is not None:
```

For Python 3.6 and above, my guess is that the tests are failing because of the threshold here. With this threshold, we end up running 47 batches. The more batches we run, the higher the likelihood of two simulations diverging due to very small differences in, e.g., implementations of basic mathematical functions. Making this threshold higher will help to reduce the number of batches. For example, picking a value of 0.0045 would result in only 12 total batches.

Reply:
OK, although I note that in the equivalent _keff_trigger_from_xml_element method, the keys (there are only two) are hard-coded. I shall do the same for _create_output_subelement which also loops over unsorted keys.